File: LeetCode/Graph/DFS/Golang/Leetcode652-find-duplicate-subtrees.go
package sbgo

func findDuplicateSubtrees(root *TreeNode) []*TreeNode {
  ans,rec,mp := []*TreeNode{},make(map[int]int),make(map[string]int)
  cnt := 0
  var dfs func(*TreeNode) int 
  dfs = func(node *TreeNode) int {
    if node == nil {
      return 0
    }
    cur := node.Val
    lhs := dfs(node.Left)
    rhs := dfs(node.Right)
    key := strconv.Itoa(cur) + " " + strconv.Itoa(lhs) + " " + strconv.Itoa(rhs)
    if _,ok := mp[key];!ok {
      cnt++
      mp[key] = cnt
    }
    id := mp[key]
    rec[id]++
    if rec[id] == 2 {
      ans = append(ans,node)
    }
    return id
  } 
  dfs(root)
  return ans
}

File: LeetCode/Graph/DFS/Python/Leetcode652-find-duplicate-subtrees.py
class Solution:
    def findDuplicateSubtrees(
            self, root: Optional[TreeNode]) -> List[Optional[TreeNode]]:
        ans, cnt, rec, mp = [], 0, defaultdict(int), defaultdict(int)

        def dfs(node):
            nonlocal cnt
            if not node: return 0
            cur = node.val
            lhs = dfs(node.left)
            rhs = dfs(node.right)
            key = str(cur) + ' ' + str(lhs) + ' ' + str(rhs)
            if mp[key] == 0:
                cnt += 1
                mp[key] = cnt
            id = mp[key]
            rec[id] += 1
            if rec[id] == 2: ans.append(node)
            return id

        dfs(root)
        return ans
